openverse_catalog/dags/providers/provider_api_scripts/finnish_museums.py

```diff
@@ -26,16 +43,71 @@ class FinnishMuseumsDataIngester(ProviderDataIngester):
     delay = 5
     format_type = "0/Image/"
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # A flag that is turned on each time we start ingestion for a new set of
+        # query params. This is useful for logging information on only the first
+        # batch of each set.
+        self.new_iteration = True
+
+        # Build the list of timestamp pairs once, so they can be reused for each
+        # building queried.
+        self.timestamp_pairs = self._get_timestamp_query_params_list(self.date)
+
+    @staticmethod
+    def _get_timestamp_query_params_list(date):
+        """
+        The Finnish Museums API behaves unexpectedly when querying large datasets,
+        resulting in large numbers of duplicates and eventual DAG timeouts
+        (see https://github.com/WordPress/openverse-catalog/pull/879 for more
+        details). To avoid this, we build a list of timestamp pairs that divide
+        the ingestion date into equal portions of the 24-hour period, and run
+        ingestion separately for each time slice.
+        """
+        # Get the logical date for the DagRun
+        utc_date = datetime.strptime(date, "%Y-%m-%d").replace(tzinfo=timezone.utc)
+
+        seconds_in_a_day = 86400
+        number_of_divisions = 48  # Half-hour increments
+        portion = int(seconds_in_a_day / number_of_divisions)
+
+        def _format_timestamp(d):
+            return d.isoformat().replace("+00:00", "Z")
+
+        # Generate the start/end timestamps for each half-hour 'slice' of the day
+        pair_list = [
+            (
+                _format_timestamp(utc_date + timedelta(seconds=i * portion)),
+                _format_timestamp(utc_date + timedelta(seconds=(i + 1) * portion)),
+            )
+            for i in range(number_of_divisions)
+        ]
+        return pair_list
+
     def ingest_records(self, **kwargs):
         for building in BUILDINGS:
             logger.info(f"Obtaining images of building {building}")
-            super().ingest_records(building=building)
+
+            for start_ts, end_ts in self.timestamp_pairs:
+                self.new_iteration = True
```

Comment:
Because this DAG is so sparse, one optimization that might be nice would be to first request the entire day's worth of data, and then only bother breaking it down into increments if the `resultCount` is above some threshold (or even just non-zero!) 🤔

Reply:
That'd be a good optimization! Maybe we could make an issue for it to include down the line?

Reply:
I agree! #913